fix(mme): PDUSessionResourceSetupReq and PDUSessionEstablishmentAccept clubbed into Single NGAP Message

ganeshg87 opened the pull request:
Test Cases:
  - verified PDUSession setup with UERANSIM

Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>

<!--
    Tag your PR title with the components that it touches along with
    the type of change
    E.g. "fix(orc8r): Fix reindexer race condition" or "feat(agw) ..."
-->

## Summary

   AMF clubs both PDUSessionResourceSetupReq and PDUSessionEstablishmentAccept into Single NGAP Message and sends to gNB during PDU Session setup.

## Test Plan

- verified PDU Session setup with UERANSIM
![image](https://user-images.githubusercontent.com/83060027/138907205-e2300cac-9db2-4b82-96a4-75cb1b959c7e.png)

[mme_pdu_accept_resource_setup.log](https://github.com/magma/magma/files/7419274/mme_pdu_accept_resource_setup.log)

## Additional Information

- [ ] This change is backwards-breaking

<!--
    If this is a backwards-breaking change, document the upgrade instructions.
    All upgrade instructions for backwards-breaking changes will be aggregated
    in the next release's changelog so this is very important to fill out.
-->

ganeshg87 committed:
fix(mme): PDUSessionResourceSetupReq and PDUSessionEstablishmentAccept
Clubbed into Single NGAP Message

Test Cases:
  - verified PDUSession setup with UERANSIM

Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>

ulaskozat commented on lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h:
```diff
@@ -785,7 +785,7 @@ void amf_smf_context_cleanup_pdu_session(ue_m5gmm_context_s* ue_context);
 int pdu_session_resource_setup_request(
     ue_m5gmm_context_s* ue_context, amf_ue_ngap_id_t amf_ue_ngap_id,
-    std::shared_ptr<smf_context_t>);
+    std::shared_ptr<smf_context_t>, bstring);
```
provide the parameter names

ganeshg87 replied:
addressed review comment, added parameter names

ulaskozat commented on lte/gateway/c/core/oai/tasks/ngap/ngap_amf_nas_procedures.c:
```diff
@@ -1201,6 +1201,19 @@ int ngap_amf_nas_pdusession_resource_setup_stream(
 
     ngap_pdusession_setup_item_ies->pDUSessionID = session_item->Pdu_Session_ID;
 
+    ngap_pdusession_setup_item_ies->pDUSessionNAS_PDU =
+        calloc(1, sizeof(Ngap_NAS_PDU_t));
+    ngap_pdusession_setup_item_ies->pDUSessionNAS_PDU->size =
+        blength(pdusession_resource_setup_req->nas_pdu);
+
+    ngap_pdusession_setup_item_ies->pDUSessionNAS_PDU->buf = calloc(
+        blength(pdusession_resource_setup_req->nas_pdu), sizeof(uint8_t));
+
+    memcpy(
+        ngap_pdusession_setup_item_ies->pDUSessionNAS_PDU->buf,
+        bdata(pdusession_resource_setup_req->nas_pdu),
+        ngap_pdusession_setup_item_ies->pDUSessionNAS_PDU->size);
+
```
where are these dynamically allocated memory got freed up?

ganeshg87 replied:
IEs allocated dynamically and added to Ngap_NGAP_PDU_t(pdu). 
all IEs memory cleared inside ngap_amf_encode_pdu API by ASN_STRUCT_FREE_CONTENTS_ONLY.

we have a UT case, it was validated memory leak(pdu_sess_resource_setup_req_sunny_day).
Please find latest  NGAP UT results, no leak reported.
![image](https://user-images.githubusercontent.com/83060027/140014426-57ac2d72-b667-44e0-b55a-f8876d8be7a9.png)

ulaskozat reviewed (changes requested):
I do not see where the allocated memory got freed.

ganeshg87 committed:
addressed review comments-R0

Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>

ulaskozat reviewed (approved):
LGTM

ulaskozat commented:
@ganeshg87 can you rebase and update as Magma-OAI-Jenkins pipeline has been failing and should be fixed now.

ganeshg87 commented:
> @ganeshg87 can you rebase and update as Magma-OAI-Jenkins pipeline has been failing and should be fixed now.
@ulaskozat, 
rebased and updated PR, CI tasks are in progress
